File: frontend/src/services/taskService.js
import apiClient from './apiClient';

/**
 * Task service for interacting with the tasks API
 */
export const taskService = {
  /**
   * Get all tasks for the current user
   * @param {Object} params - Query parameters
   * @param {number} params.skip - Number of items to skip
   * @param {number} params.limit - Number of items to return
   * @param {string} params.status - Filter by status
   * @returns {Promise} Promise resolving to tasks array
   */
  getTasks: async (params = {}) => {
    try {
      const response = await apiClient.get('/tasks/', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching tasks:', error);
      throw error;
    }
  },

  /**
   * Get a specific task by ID
   * @param {string} taskId - Task ID
   * @returns {Promise} Promise resolving to task object
   */
  getTask: async (taskId) => {
    try {
      const response = await apiClient.get(`/tasks/${taskId}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching task ${taskId}:`, error);
      throw error;
    }
  },

  /**
   * Create a new task
   * @param {Object} taskData - Task data
   * @returns {Promise} Promise resolving to created task
   */
  createTask: async (taskData) => {
    try {
      const response = await apiClient.post('/tasks/', taskData);
      return response.data;
    } catch (error) {
      console.error('Error creating task:', error);
      throw error;
    }
  },

  /**
   * Update an existing task
   * @param {string} taskId - Task ID
   * @param {Object} taskData - Updated task data
   * @returns {Promise} Promise resolving to updated task
   */
  updateTask: async (taskId, taskData) => {
    try {
      const response = await apiClient.put(`/tasks/${taskId}`, taskData);
      return response.data;
    } catch (error) {
      console.error(`Error updating task ${taskId}:`, error);
      throw error;
    }
  },

  /**
   * Delete a task
   * @param {string} taskId - Task ID
   * @returns {Promise} Promise resolving when task is deleted
   */
  deleteTask: async (taskId) => {
    try {
      await apiClient.delete(`/tasks/${taskId}`);
      return true;
    } catch (error) {
      console.error(`Error deleting task ${taskId}:`, error);
      throw error;
    }
  },

  /**
   * Mark a task as completed
   * @param {string} taskId - Task ID
   * @param {number} actualDuration - Actual duration in minutes
   * @returns {Promise} Promise resolving to updated task
   */
  completeTask: async (taskId, actualDuration) => {
    try {
      const response = await apiClient.post(`/tasks/${taskId}/complete`, { actual_duration: actualDuration });
      return response.data;
    } catch (error) {
      console.error(`Error completing task ${taskId}:`, error);
      throw error;
    }
  }
};

export default taskService;

File: backend/app/services/task_service.py
from sqlalchemy.orm import Session
from uuid import UUID
from datetime import datetime
from typing import List, Optional

from app.models.task import Task, TaskCreate, TaskUpdate, TaskStatus

class TaskService:
    def __init__(self, db: Session):
        self.db = db
        
    def create_task(self, user_id: UUID, task: TaskCreate) -> Task:
        """Create a new task for a user."""
        db_task = Task(
            **task.dict(),
            user_id=user_id,
            created_at=datetime.now(),
            updated_at=datetime.now()
        )
        self.db.add(db_task)
        self.db.commit()
        self.db.refresh(db_task)
        return db_task
        
    def get_tasks(self, user_id: UUID, skip: int = 0, limit: int = 100, status: Optional[str] = None) -> List[Task]:
        """Get all tasks for a user, optionally filtered by status."""
        query = self.db.query(Task).filter(Task.user_id == user_id)
        
        if status:
            query = query.filter(Task.status == status)
            
        return query.order_by(Task.created_at.desc()).offset(skip).limit(limit).all()
        
    def get_task(self, user_id: UUID, task_id: UUID) -> Task:
        """Get a specific task by ID."""
        task = self.db.query(Task).filter(Task.id == task_id, Task.user_id == user_id).first()
        if not task:
            raise ValueError("Task not found")
        return task
        
    def update_task(self, user_id: UUID, task_id: UUID, task_update: TaskUpdate) -> Task:
        """Update a task."""
        task = self.get_task(user_id, task_id)
        
        update_data = task_update.dict(exclude_unset=True)
        for key, value in update_data.items():
            setattr(task, key, value)
            
        task.updated_at = datetime.now()
        self.db.commit()
        self.db.refresh(task)
        return task
        
    def delete_task(self, user_id: UUID, task_id: UUID) -> None:
        """Delete a task."""
        task = self.get_task(user_id, task_id)
        self.db.delete(task)
        self.db.commit()
        
    def complete_task(self, user_id: UUID, task_id: UUID, actual_duration: Optional[int] = None) -> Task:
        """Mark a task as completed."""
        task = self.get_task(user_id, task_id)
        task.status = TaskStatus.completed
        task.completed_at = datetime.now()
        task.actual_duration = actual_duration
        task.updated_at = datetime.now()
        self.db.commit()
        self.db.refresh(task)
        return task
